### DISCUSSION TITLE
List of external services in Console for every project

### DISCUSSION THREAD

**Discussion by lucascanna** [mia-platform/community#108] [Ideas]
Hi guys!

In enterprise environments, supporting the IT governance is one of the main values of Mia-Platform Console.

A very common problem in these environments is to identify, for each project, which are the external dependencies, i.e. the external services (databases, HTTP APIs, FTP servers etc.) that the microservices of the project need to work properly.

For example, this list is essential in cluster migrations, security audits, systems decommissioning.

Usually, the URLs to external systems are environment variables of the microservices.
Therefore, there are two possible solutions to have this list in the Console:

1. Trusting the user: ask the user to label environment variables as "external URL" if they are URLs to external dependencies
2. Automatic label: mark the environment variable as "external URL" if it matches the URL pattern

I would like to hear from you what do you think about this possible feature.
Thanks!

**Comment by davidebianchi** [comment]
Hi @Lucascanna! To have all the resources which interact with the platform, we are thinking about many possibilities. 
All of this integrations are based on the project or clusters providers (now we already have some providers, for example git provider, CI/CD provider, k8s cluster), so you could add an external service as a provider and connect with a service.

If I take mongo as an example, if we will support mongo atlas as a provider, maybe in the future it would be possible to generate the username and password required to connect to mongo and connect directly the service in this way, generating a new secret instead of use the variable interpolation as nowadays.
Another example could be done also with redis, s3, gcs, and so on all the services which it is important to know for a project.
This should already be in the future roadmap ideas, but we should improve it before to implement.

Can this be a possible solution to improve the governance for your point of view?

**Reply by lucascanna** [reply]
Hi @davidebianchi!

Yes, the solution with providers could be good but, in my opinion, we need to force somehow developers to use providers if they want their microservices to communicate with external systems. It should be the only possible way to do it. 
These enforcement would enable a lot of other interesting features like checking, at deploy time, the reachability of all the providers of the project or read in every moment the list of dependencies of a project.

**Reply by davidebianchi** [reply]
Console should not force you to follow a specific path, but it could be enforced by the company maybe.
But if there is a provider already listed and easy to manage, why an user should use it in another way? Force to use only supported providers could led you to have some friction using unsupported provider types.

**Reply by lucascanna** [reply]
Well, I agree in making the company choose whether providers are mandatory or not. 

However, my request was not only related to market tools like MongoDB, Redis, S3 etc. which are usually well-known at infrastructure level and so you could add them all in the company configuration phase.   

I am thinking also about custom services, maybe deployed somewhere outside the K8S cluster. For example, imagine to have a certain team which develops a software deployed on an on-premise VM. At some point, a project managed with Mia-Platform Console needs to interact with that service. Then, the developer would have two choices:

1. create the provider from console (or ask someone with right permissions to create it) and then use it from the service he is implementing
2. create public or secreted environment variables and use it in the service he is implementing

I think that developers would be really tempted to use solution 2 as it is faster.

**Reply by davidebianchi** [reply]
Ok clear, I agree with you!

For this, it's maybe possible to save an external proxy in the marketplace and use it inside the project. This could be an extension of this feature maybe to optimize the op to contact the service. 

Because yes, the solution number two is faster, but what if the url of the service change, or to have the governance of when it is configured to use this service?
And I think the choice is all about how it is easy to configure it. For example, if there are already all the services of the company visible from the Console, it will be so easy to add it directly instead of manually search the IP/host and the authentication of the service.

If there is a need for something like API key from the external service for every project, or a client credentials, it could be done automatically. And at this point maybe have the auto completion inside the Console will help to improve the dev experience.

**Comment by theliuk** [comment]
We are actually working on this our P4SaMD product. We won't to identify all the so called [SOUP](https://en.wikipedia.org/wiki/Software_of_unknown_pedigree) because they must be validated. 

In the context of the console, such dependencies are:

- Docker images of marketplace components or Docker images found on the web
- External APIs (not only REST HTTP)
- External libraries

We will show the user all of those dependencies to be validated

**Comment by akelity** [comment]
This feature has been positively evaluated and may be considered for inclusion in the roadmap in the future, depending on priorities and overall specific needs.
